plugins/aggregators/derivative/derivative.go

```diff
@@ -0,0 +1,209 @@
+package derivative
+
+import (
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/aggregators"
+)
+
+type Derivative struct {
+	Variable    string
+	Infix       string
+	MaxRollOver uint
+	cache       map[uint64]aggregate
+}
+
+type aggregate struct {
+	first    *event
+	last     *event
+	name     string
+	tags     map[string]string
+	rollOver uint
+}
+
+type event struct {
+	fields map[string]float64
+	time   time.Time
+}
+
+func NewDerivative() telegraf.Aggregator {
+	derivative := &Derivative{Infix: "_by_", MaxRollOver: 10}
+	derivative.cache = make(map[uint64]aggregate)
+	derivative.Reset()
+	return derivative
+}
+
+var sampleConfig = `
+  ## This Aggregator will estimate a derivative for each field, which is
+	## contained in both the first and last metric of the aggregation interval.
+  ## Without further configuration the derivative will be calculated with
+	## respect to the time difference between these two measurements in seconds.
+	## The formula applied is for every field:
+	##
+	##               value_last - value_first
+	## derivative = --------------------------
+  ##              time_difference_in_seconds
+  ##
+	## The resulting derivative will be named *fieldname_by_seconds*. The infix
+	## "_by_" can be configured by the *infix* parameter.
+	# infix = "_wrt_"
+	##
+	## As an abstraction the derivative can be calculated not only by the time
+	## difference but by the difference of a field, which is contained in the
+	## measurement. This field is assumed to be monotonously increasing. This
+	## feature is used by specifying a *variable*.
+	# variable = "parameter"
+	##
+	## When using a field as the derivation parameter the name of that field will
+	## be used for the resulting derivative, e.g. *fieldname_by_parameter*.
+	##
+	## As a genereal parameter the period needs to be given. It determines the
+	## period for which a derivative will be calculated
+	period = "30s"
+	##
+	## Note, that the calculation is based on the actual timestamp of the
+	## measurements. When there is only one measurement during that period, the
+	## measurement will be rolled over to the next period. The maximum number of
+	## such roll-overs can be configured with a default of 10.
+	# max_roll_over = 10
+	##
+`
+
+func (d *Derivative) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *Derivative) Description() string {
+	return "Calculates a derivative for every field."
+}
+
+func (d *Derivative) Add(in telegraf.Metric) {
+	id := in.HashID()
+	if current, ok := d.cache[id]; !ok {
+		// hit an uncached metric, create caches for first time:
+		d.cache[id] = newAggregate(in)
+	} else {
+		if current.first.time.After(in.Time()) {
+			current.first = newEvent(in)
+			current.rollOver = 0
+		}
+		if current.first.time.Equal(in.Time()) {
+			current.first.fields = upsertConvertedFields(in.Fields(), current.first.fields)
+			current.rollOver = 0
+		}
```

Comment:
I don't understand why you handle that differently... You always want to replace the field values or am I wrong here?

Reply:
I see. Now imagine you have two such metrics leading to the same ID having the same timestamp but two field-sets, say the first on has `in` the second one `out`. Further assume a first `period` where those metric arrived you have a metric with `in, out` in `cache`, nice.
Now after the roll-over for some reason only the first metric arrived within `period` with a timestamp *earlier* than `first`. This time you will keep the replace `first with a metric having **only** the `in` field and you drop the `out` field... Is this really what you are intending?

Reply:
Furthermore, while writing this up I recognized another (potential) issue... Imaging the metrics that enter the plugin are strictly ordered by time (e.g. slow query intervals with only one plugin). In these cases the newly arriving metrics are **always later** than `first` even after a roll-over. In this scenario `first` stays the same until `MaxRollover` is reached!!! I guess you need to additionally check if the newly arrived metric is the *first* in the `period` and replace `first` in this case. You can use `current.rollOver > 0` to distinguish the two cases IMO.

Reply:
To the question on dropping fields when replacing `first` event: There are two cases: derivatives by timestamp (default) and derivatives by variables (another field). In both cases receiving the earlier metric with only `in` can still be followed by a metric adding `out` for this earlier timestamp. I cannot know this at that time. I see the following implications:

**derivatives by timestamp**
In this case, the correctness of the calculation depends on the field values at the timestamp used in the denominator. If I replace the `first` event by an earlier event, I can only use the fields present for that timestamp and hope there will be an additional event adding the missing metrics. In you example, I would hope that there would be an update for `out` incoming. Otherwise I would rather not to emit a value for `out` than to emit a false one.

**derivatives by variables (other fields)**
Here it is essential, that the field for which the derivative is calculated and the variable in the denominator are correlated to the same event. Otherwise the calculation result would be wrong. Going with your example, calculus give us the following formula:

```
 d(in)    d(in) / dt
------ = -----------
d(out)   d(out) / dt
```

This relation is once again only true, if `in` and `out` are taken from the same timestamp. Once again I went for not emitting a value rather that a wrong value.

So yes, in both cases dropping the field was intentional. Did I miss something in this?

Reply:
Your scenario with the the slow incoming metrics is a flaw in the `cache` clean-up. Elements are only removed from the `cache` when `MaxRollOver` is reached. Naively I would think, that it should also be removed, when a derivative is emitted in `Push`. I will look at the tests and perhaps add some examples to outline the desired behaviour. If there are changes required, I will make them in the code.

Reply:
> So yes, in both cases dropping the field was intentional. Did I miss something in this?

No I don't think you missed something, but the problem is that in the described cases you potentially drop fields and/or, if I interpret your code correctly, fail calculating the denominator in `calculateDenominator()`
```
￼	if len(d.Variable) != 0 {
￼		return aggregate.last.fields[d.Variable] - aggregate.first.fields[d.Variable]
￼	}
￼}
```
if for any reason `d.Variable` did not arrive. The lookup will then be `nil`.
Even if you go with the time denominator, the problem is that the user will end up with `null` field values in the database for the missing fields and has no indication of why. But I think to fix this, you would need to completely restructure the code, keeping track of the individual fields instead of metrics...
So to cut a long story short: Keep what you have, but please fix `calculateDenominator()` for cases where `d.Variable` did not arrive for any of the metrics.

Reply:
> Your scenario with the the slow incoming metrics is a flaw in the `cache` clean-up. Elements are only removed from the `cache` when `MaxRollOver` is reached. Naively I would think, that it should also be removed, when a derivative is emitted in `Push`. I will look at the tests and perhaps add some examples to outline the desired behaviour. If there are changes required, I will make them in the code.

I don't think it is about cleaning the `cache` but rather about the logic of replacing the values. I will outline 4 arriving samples with the form `timestamp: value` (imaging a metric with only one field). Say we get `0: 1.0; 1: 2.0; 2: 1.5; 3: 2.2` and compute the derivative by time. With your code you get
| incoming | cache first (start) | cache last (start)  | rollover after `Reset()` | cache first (end) | cache last (end) | outgoing | comment |
|---|---|---|---|---|---|---|---|
|0: 1.0| - | - | - | 0: 1.0 |  0: 1.0 | - | denominator zero |
|1: 2.0| 0: 1.0 | 0: 1.0 | 1 | 0: 1.0 |  1: 2.0 | 1.0 | so far so good, `rollover` is reset to zero |
|2: 1.5; 3: 2.2| 1: 2.0 | 1: 2.0 | 1 | 1: 2.0 |  3: 2.2 | 0.1 | `current.first.time` is neither after nor equal `in.Time()` for the first arriving metric|

Is this what you want? For the last line in the table I could expect
| incoming | cache first (start) | cache last (start)  | rollover after `Reset()` | cache first (end) | cache last (end) | outgoing | comment |
|---|---|---|---|---|---|---|---|
|2: 1.5; 3: 2.2| 1: 2.0 | 1: 2.0 | 1 | 2: 1.5 |  3: 2.2 | 0.7 | |

In case multiple metrics arrive in the `period` I would expect the first one in `period` replaces `current.first`...

Reply:
Actually 0.1 is exactly what I want. This way the derivative is continuously calculated, leaving no holes in the time series. In your example taking (2: 1.5) as first value of the period would omit the decline between (1: 2.0) and (2: 1.5). This would give the false impression, that the value was continuously increasing, when in fact, there was a small decrease. You can get your desired behaviour by configuring `max_roll_over` to 0 and choosing a period, that ensures multiple values are aggregated. Dealing with jumps between aggregation periods was one of the reasons, I implemented the roll-over.

Reply:
Ok, I'm also fine with your interpretation, but can you please describe this a bit more in the README!? Feel free to use my example without the internal fields and show what happens for the different patterns (first arrival skipped, single arrival per period, no arrival per period, multiple arrivals per period....). Maybe also use a table or similar showing the different cases and describe what will be the output and give a short explanation of the "general rule".

Reply:
I will do this. It also noticed, that the behaviour is not straight-forward. But it is useful.